proposed/2023/release-staging-and-deprecation.md

```diff
@@ -0,0 +1,90 @@
+
+## Motivation 
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+Many products (Apps / Tools / Frameworks / Utilities / etc. ...) consist of multiple interdependent packages which should be published together as a single set. Currently, NuGet provides no functionality for definding / grouping packages into such sets. This results in package owners having to keep track of these package sets on their own. There are three specific sets of scenarios which this feature will simplify dramatically:
+1. When preparing a release, product owners would like to have a mechanism to pre-stage their product packages, so that the release will consist only of a single action of marking the set of packages as `published`, which would result all the packages in the set to be marked as published at the same time.
+2. Sometimes developers have to take a dependency on several related packages in their projects, which are part of the same "release". Without having a concept of the `release` developers have to guess which versions of those packages are compatible. As a result, most of hte time developers would simply pick the latest version of all these packages, which somtimes can lead to incompatibility issues down the road.
+3. Keeping a package healthy on NuGet also means keeping the list of packages aligned with it support policy. Hence, package owners try to deprecate packages which are out of support. Without ability to group packages, package owners end up keeping track of all the packages which belong to the same release and run the deprecation over a discrete set of packages when necessary. Being able to group packages together would enable deprecating a whole release at once, knowing that all the packages which belong to that release will get deprecated.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+There is a new field in the package metadata named `release`, which indicates the unique release of the product the package belongs to.
+Server implementations are expected to treat each `release` as scoped to an account or similar concept, depending on the server's implementation details.
+For example, on nuget.org, User1 and User2 can both use a `release` named `V1.0.0`, but when User1 operates on their `V1.0.0` release, it does not affect User2's release with the same name.
+
+When a package is being built, that field in the metadata is set to the same value for all the related packages, that are built together.
+
+Now, with this information in place, new set of NuGet APIs / extensions allow the user to run the following actions:
+- **Stage** a package for later publishing
```

Comment:
What is an acceptable level of visibility for staged packages? I am exploring this because it has real impact on implementation cost.

Suppose a staged package is simply unlisted. It's findable by an unauthenticated user who knows the version number but does not appear in package search or version lists in VS or NuGet.org UI. This is a cheaper thing for NuGet.org to implement. Publishing the release/staged packages would simply be marking N packages as listed.

If we need to do a mega, big bang release of hundreds of totally unpublished, staged packages, this can impact the availability of our publishing pipeline. Today, there is a relationship between package push rate limits and throughput capabilities of our publishing pipeline, i.e. we picked 350 pushes per hour based on historical data and load testing of our publishing pipeline.

If you expect a staged release to do all of the publish at once, and go from completely invisible packages to normally published immediately, this introduces two challenges to the package source:
1. You need to transactionally move N packages from "staged" to "available"
2. This eliminates much of the usefulness for upload throttling since you can upload to a stage as slowly as you want, building a massive queue of packages, and the publish in one moment creating a huge spike in real publishes at that moment.

There are middle grounds where the package is available on the [Package content resource](https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource) but not the [Package metadata resource](https://learn.microsoft.com/en-us/nuget/api/registration-base-url-resource) which could perhaps give NuGet.org's publishing pipeline more breathing room, but that all would be costly to investigate. 

I am assuming here that staged packages go through our asynchronous validation pipeline (malware scanning, repository signing) prior to release. This would be necessary for a timely release since you don't want one package stuck in a slower malware scan to block or break the release. I have no concerns for this from a scalability perspective since our validation pipeline is typically not our bottleneck. It's the publishing pipeline that updates the various V3 API resources.

Reply:
I think I've covered this by the Asynchronous operations behavior mentioned in the proposal.